### PR TITLE
Improve spawn-daemon and Nomad Client usage of it

### PR DESCRIPTION
**Pull request by dadgar**
This PR does the following:

* spawn-daemon is refactored, updated to be cross-platform and saves the exit code to disk.
* Create Spawner package that is used to create the spawn-daemon and handles IPC with it.
* exec is disabled on all platforms other than linux running as root (only way to do isolation).
* exec_linux is updated to assume cgroups, support systemd, actually implement the Open() interface.
* Create a BasicExecutor which can be used by both exec_universal and raw_exec.

### CODE REVIEW

File: client/spawn/spawn.go [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+	// If there is data it means that the file has already been written.
+	if stat.Size() > 0 {
+		return s.readExitCode()
+	}
+
+	// Wait on watcher.
+	for {
+		select {
+		case event := <-watcher.Events:
```
Comment (diptanu) [inline client/spawn/spawn.go]: Are we watching for events to the file where spawn daemon is going to write the exit code for the process it starts using inotify and also checking whether the spawn daemon is alive every 5s at the same time?

If we are checking if spawn daemon is alive every 5s, then we could just read the file for the exit code when spwan daemon has exited.

**Comment by diptanu** [comment]
LGTM overall. I have a question regarding our package structure. Since the executor package is used from within the drivers and not exposed out side the drivers, can we move the package from client/ to client/drive/executor? This implies that the executor package exposes some APIs which helps to write drivers to run commands with isolation.

File: client/spawn/spawn.go (hunk 1 above)
Comment (dadgar) [inline client/spawn/spawn.go]: Good catch!

**Comment by dadgar** [comment]
Thanks for the review @diptanu. Lets do the package structure changes in another PR

**Comment by catsby** [comment]
My `git-bisect` foo is weak, but there was a regression in the `TestExecDriver_Start_Artifact_basic` test and `git-bisect` suggested it was commit 3576f489932ede9b12640d5bbfbf44324f0cc527 (or around there I guess). 

In a Linux VM, as root, this will timeout:

```
$ go test -v ./client/driver -run="TestExecDriver_Start_Artifact_basic"
```

Extending [this timeout][1] to `15` gives this error:

```console
root@nomad:/opt/gopath/src/github.com/hashicorp/nomad# go test -v ./client/driver -run="TestExecDriver_Start_Artifact_basic"
=== RUN   TestExecDriver_Start_Artifact_basic
--- FAIL: TestExecDriver_Start_Artifact_basic (12.27s)
	exec_test.go:159: err: 2 error(s) occurred:

		* Empty state file: /tmp/54bd7fc0-31eb-3179-378c-4b178d07525c/sleep_exit_status
		* Task exited with code: -1
FAIL
exit status 1
FAIL	github.com/hashicorp/nomad/client/driver	12.277s
```


[1]: https://github.com/hashicorp/nomad/blob/master/client/driver/exec_test.go#L161

**Comment by catsby** [comment]
@dadgar pointed out I need to do `make dev`, which resolved the issue. Sorry for the noise!

**Pull request merged**